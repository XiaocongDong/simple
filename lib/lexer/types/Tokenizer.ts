import { TOKEN_TYPE } from "./token"

export type IState = string
export type ITokenTypeGenerator = (value: string) => TOKEN_TYPE
export type IStateConfig = {
  isEnd: boolean
  TokenType?: TOKEN_TYPE|ITokenTypeGenerator
  transitions?: Array<ITransition>
}
export type IStatesConfig = Record<IState, IStateConfig>

interface ITransition {
  checker: RegExp|string|((ch: string) => boolean)
  state: string
}

export type IConfig = {
  states: IStatesConfig
  initialState: IState
}
