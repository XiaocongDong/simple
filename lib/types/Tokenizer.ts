import { TOKEN_TYPE } from "./token"

export type IState = string
export type ITokenTypeGenerator = (value: string) => TOKEN_TYPE
export type IStateConfig = {
  isEnd: boolean
  TokenType?: TOKEN_TYPE|ITokenTypeGenerator
  transitions?: Array<ITransition>
}

interface ITransition {
  checker: RegExp|string|((ch: string) => boolean)
  state: string
}

export type IConfig = {
  states: Record<IState, IStateConfig>
  initialState: IState
}
