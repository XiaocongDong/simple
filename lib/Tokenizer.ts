import Tokenizer from './lexer/Tokenizer'
import config from './config/Tokenizer'

const tokenizer = new Tokenizer(config)

export default tokenizer
