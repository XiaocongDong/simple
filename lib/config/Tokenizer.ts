import { IConfig } from "../lexer/types/Tokenizer";
import { TOKEN_TYPE } from "../lexer/types/token";
import { LET, CONST, BREAK, CONTINUE, IF, ELSE, WHILE, FUNCTION, UNDEFINED, NULL, TRUE, FALSE, NEW } from "../constants";

enum State {
  INITIAL = 'INITIAL',
  NUMBER_LITERAL = 'NUMBER_LITERAL',
  START_IDENTIFIER = 'IDENTIFER',
  IDENTIFER = 'IDENTIFER',
  START_STRING_LITERAL = 'START_STRING_LITERAL',
  STRING_LITERAL = 'STRING_LITERAL',

  PLUS = 'PLUS',
  PLUS_PLUS = 'PLUS_PLUS',
  MINUS = 'MINUS',
  MINUS_MINUS = 'MINUS_MINUS',
  MULTIPLY = 'MULTIPLY',
  DIVIDE = 'DIVIDE',
  PLUS_ASSIGN = 'PLUS_ASSIGN',
  MINUS_ASSIGN = 'MINUS_ASSIGN',
  MULTIPLY_ASSIGN = 'MULTIPLY_ASSIGN',
  DIVIDE_ASSIGN = 'DIVIDE_ASSIGN',
  LESS_THAN = 'LESS_THAN',
  LESS_EQUAL_THAN = 'LESS_EQUAL_THAN',
  GREATER_THAN = 'GREATER_THAN',
  GREATER_EQUAL_THAN = 'GREATER_EQUAL_THAN',

  ASSIGN = 'ASSIGN',
  COMMA = 'COMMA',
  DOT = 'DOT',

  LEFT_PAREN = 'LEFT_PAREN',
  RIGHT_PAREN = 'RIGHT_PAREN',
  SEMI_COLON = 'SEMI_COLON',
  LEFT_CURLY_BRACE = 'LEFT_CURLY_BRACE',
  RIGHT_CURLY_BRACE = 'RIGHT_CURLY_BRACE',
  LEFT_SQUARE_BRACKET = 'LEFT_SQUARE_BRACKET',
  RIGHT_SQUARE_BRACKET = 'RIGHT_SQUARE_BRACKET',
  COLON = 'COLON',

  START_OR = 'START_OR',
  EQUAL = 'EQUAL',
  OR = 'OR',
  START_AND = 'START_AND',
  AND = 'AND'
}

const keywordTokenTypeLookup: {[k: string]: TOKEN_TYPE} = {
  [LET]: TOKEN_TYPE.LET,
  [CONST]: TOKEN_TYPE.CONST,
  [BREAK]: TOKEN_TYPE.BREAK,
  [CONTINUE]: TOKEN_TYPE.CONTINUE,
  [IF]: TOKEN_TYPE.IF,
  [ELSE]: TOKEN_TYPE.ELSE,
  [WHILE]: TOKEN_TYPE.WHILE,
  [FUNCTION]: TOKEN_TYPE.FUNCTION,
  [UNDEFINED]: TOKEN_TYPE.UNDEFINED,
  [NULL]: TOKEN_TYPE.NULL,
  [TRUE]: TOKEN_TYPE.TRUE,
  [FALSE]: TOKEN_TYPE.FALSE,
  [NEW]: TOKEN_TYPE.NEW
}

const keywordChecker = (value) => {
  if (value === 'constructor') {
    // object has an attribute named constructor in js...
    return TOKEN_TYPE.IDENTIFER
  }

  const tokenType = keywordTokenTypeLookup[value]
  return tokenType || TOKEN_TYPE.IDENTIFER
}

const config: IConfig = {
  initialState: State.INITIAL,
  states: {
    [State.INITIAL]: {
      isEnd: false,
      transitions: [
        {
          state: State.NUMBER_LITERAL,
          checker: /[0-9]/
        },
        {
          state: State.START_IDENTIFIER,
          checker: /[a-zA-Z_]/
        },
        {
          state: State.START_STRING_LITERAL,
          checker: "'"
        },
        {
          state: State.PLUS,
          checker: "+"
        },
        {
          state: State.MINUS,
          checker: "-"
        },
        {
          state: State.MULTIPLY,
          checker: "*"
        },
        {
          state: State.DIVIDE,
          checker: "/"
        },
        {
          state: State.ASSIGN,
          checker: "="
        },
        {
          state: State.START_OR,
          checker: "|"
        },
        {
          state: State.START_AND,
          checker: "&"
        },
        {
          state: State.COLON,
          checker: ':'
        },
        {
          state: State.SEMI_COLON,
          checker: ';'
        },
        {
          state: State.LEFT_PAREN,
          checker: '('
        },
        {
          state: State.RIGHT_PAREN,
          checker: ')'
        },
        {
          state: State.LEFT_CURLY_BRACE,
          checker: '{'
        },
        {
          state: State.RIGHT_CURLY_BRACE,
          checker: '}'
        },
        {
          state: State.COMMA,
          checker: ','
        },
        {
          state: State.DOT,
          checker: '.'
        },
        {
          state: State.LEFT_SQUARE_BRACKET,
          checker: '['
        },
        {
          state: State.RIGHT_SQUARE_BRACKET,
          checker: ']'
        },
        {
          state: State.LESS_THAN,
          checker: '<'
        },
       {
         state: State.GREATER_THAN,
         checker: '>'
       }
      ]
    },
    [State.NUMBER_LITERAL]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.NUMBER_LITERAL,
      transitions: [
        {
          state: State.NUMBER_LITERAL,
          checker: /[0-9\.]/
        }
      ]
    },
    [State.START_IDENTIFIER]: {
      isEnd: true,
      TokenType: keywordChecker,
      transitions: [
        {
          state: State.IDENTIFER,
          checker: /[0-9a-zA-Z_]/
        }
      ]
    },
    [State.IDENTIFER]: {
      isEnd: true,
      TokenType: keywordChecker,
      transitions: [
        {
          state: State.IDENTIFER,
          checker: /[0-9a-zA-Z_]/
        }
      ]
    },
    [State.START_STRING_LITERAL]: {
      isEnd: false,
      transitions: [
        {
          state: State.START_STRING_LITERAL,
          checker: /[^']/
        },
        {
          state: State.STRING_LITERAL,
          checker: "'"
        }
      ]
    },
    [State.STRING_LITERAL]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.STRING_LITERAL
    },
    [State.PLUS]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.PLUS,
      transitions: [
        {
          state: State.PLUS_PLUS,
          checker: "+"
        },
        {
          state: State.PLUS_ASSIGN,
          checker: "="
        }
      ]
    },
    [State.PLUS_PLUS]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.PLUS_PLUS
    },
    [State.MINUS]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.MINUS,
      transitions: [
        {
          state: State.MINUS_MINUS,
          checker: "-"
        },
        {
          state: State.MINUS_ASSIGN,
          checker: "="
        }
      ]
    },
    [State.MINUS_MINUS]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.MINUS_MINUS
    },
    [State.MINUS_ASSIGN]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.MINUS_ASSIGN
    },
    [State.MULTIPLY]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.MULTIPLY,
      transitions: [
        {
          state: State.MULTIPLY_ASSIGN,
          checker: "="
        }
      ]
    },
    [State.MULTIPLY_ASSIGN]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.MULTIPLY_ASSIGN
    },
    [State.DIVIDE]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.DIVIDE,
      transitions: [
        {
          state: State.DIVIDE_ASSIGN,
          checker: "="
        }
      ]
    },
    [State.DIVIDE_ASSIGN]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.DIVIDE_ASSIGN
    },
    [State.ASSIGN]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.ASSIGN,
      transitions: [
        {
          state: State.EQUAL,
          checker: "="
        }
      ]
    },
    [State.EQUAL]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.EQUAL
    },
    [State.START_OR]: {
      isEnd: false,
      transitions: [
        {
          state: State.OR,
          checker: "|"
        }
      ]
    },
    [State.OR]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.OR
    },
    [State.START_AND]: {
      isEnd: false,
      transitions: [
        {
          state: State.AND,
          checker: "&"
        }
      ]
    },
    [State.AND]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.AND
    },
    [State.SEMI_COLON]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.SEMI_COLON
    },
    [State.LEFT_PAREN]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.LEFT_PAREN
    },
    [State.RIGHT_PAREN]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.RIGHT_PAREN
    },
    [State.LEFT_CURLY_BRACE]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.LEFT_CURLY_BRACE
    },
    [State.RIGHT_CURLY_BRACE]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.RIGHT_CURLY_BRACE
    },
    [State.COMMA]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.COMMA
    },
    [State.DOT]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.DOT
    },
    [State.LEFT_SQUARE_BRACKET]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.LEFT_SQUARE_BRACKET
    },
    [State.RIGHT_SQUARE_BRACKET]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.RIGHT_SQUARE_BRACKET
    },
    [State.COLON]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.COLON
    },
    [State.LESS_THAN]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.LESS_THAN,
      transitions: [
        {
          checker: '=',
          state: State.LESS_EQUAL_THAN
        }
      ]
    },
    [State.LESS_EQUAL_THAN]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.LESS_EQUAL_THAN,
    },
    [State.GREATER_THAN]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.GREATER_THAN,
      transitions: [
        {
          checker: '=',
          state: TOKEN_TYPE.GREATER_EQUAL_THAN
        }
      ]
    },
    [State.GREATER_EQUAL_THAN]: {
      isEnd: true,
      TokenType: TOKEN_TYPE.GREATER_EQUAL_THAN
    }
  }
}

export default config
