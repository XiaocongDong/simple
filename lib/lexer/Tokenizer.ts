import TokenBuffer from "./TokenBuffer"
import LocationKeeper from './LocationKeeper'
import SyntaxError from '../errors/Syntax'

import { IConfig, IState, IStateConfig, ITokenTypeGenerator, IStatesConfig } from "./types/Tokenizer"
import { SPACE, NEW_LINE } from "../constants"
import { TOKEN_TYPE, IToken } from "./types/token"

class Tokenizer {
  private tokenBuffer: TokenBuffer = new TokenBuffer()
  private statesConfig: IStatesConfig = null
  private initialState: IState = null
  private state: IState = null
  private locationKeeper: LocationKeeper = new LocationKeeper()
  private buffer: string = ''
  
  constructor(config: IConfig) {
    this.statesConfig = config.states
    this.initialState = config.initialState
    this.state = config.initialState
  }

  getTokenBuffer() {
    return this.tokenBuffer
  }

  parse(code: string) {
    for(let i = 0; i < code.length; i++) {
      const ch = code[i]
      this.consume(ch)
    }
    this.end()
  }

  end() {
    if (this.state === this.initialState) {
      if (this.buffer) {
        throw new SyntaxError('Unexpected EOF', this.locationKeeper.getCurrentLocation())
      }
      return
    }
    
    const currentStateConfig: IStateConfig = this.statesConfig[this.state]
    if (currentStateConfig.isEnd) {
      this.addToken(currentStateConfig.TokenType)
    }
  }

  reset() {
    this.state = this.initialState
    this.buffer = ''
  }

  addToken(tokenType: TOKEN_TYPE|ITokenTypeGenerator) {
    const startLocation = this.locationKeeper.getPrevLocation()
    const endLocation = this.locationKeeper.getCurrentLocation()
    const token: IToken = {
      type: (typeof tokenType === 'function' ? tokenType(this.buffer) : tokenType),
      value: this.buffer,
      range: {
        start: startLocation,
        end: endLocation
      }
    }
    this.tokenBuffer.write(token)
  }

  consume(ch: string) {
    if ((ch === SPACE) && this.state === this.initialState) {
      this.locationKeeper.consume(ch)
      return
    }

    const currentStateConfig: IStateConfig = this.statesConfig[this.state]
    if (!currentStateConfig) {
      throw new Error(`Missing state config for ${this.state}`)
    }

    const transitions = currentStateConfig.transitions
    if (!transitions) {
      if (currentStateConfig.isEnd) {
        this.addToken(currentStateConfig.TokenType)
        this.reset()
        this.consume(ch)
        return
      }

      throw new SyntaxError(`Unexpected character ${ch}`, this.locationKeeper.getCurrentLocation())
    }
  
    const targetTransition = transitions.find(({ checker }) => {
      if (typeof checker === 'string') {
        return ch === checker
      }

      if (checker instanceof RegExp) {
        return checker.test(ch)
      }

      return checker(ch)
    })

    if (!targetTransition) {
      if (currentStateConfig.isEnd) {
        this.addToken(currentStateConfig.TokenType)
        this.reset()
        this.consume(ch)
        return
      }

      this.locationKeeper.consume(ch)
      throw new SyntaxError('Invalid or unexpected token', this.locationKeeper.getCurrentLocation())      
    }

    this.locationKeeper.consume(ch)
    // when state start to transfer from initial state to other state, mark current location
    if (this.state === this.initialState && targetTransition.state !== this.initialState) {
      this.locationKeeper.markLocation()
    }

    this.state = targetTransition.state
    this.buffer += ch
  }
}

export default Tokenizer
