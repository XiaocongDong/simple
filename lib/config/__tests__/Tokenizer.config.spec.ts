import config from '../Tokenizer'
import Tokenizer from '../../lexer/Tokenizer'

const testCases: Array<{code: string, description: string}> = [
  {
    code:
    `let f = function(x, y, z) {
      return x+y*z
    };
    const a = [1, 2, 3, {1, 2, 3}, 'hi'];
    const o = {
      name: 'sean',
      age: 27
    };
    `,
    description: 'it should generate correct buffer'
  },
  {
    code: `
      let a = 10; // a is not defined
      let b = 40; // b is not defined
    `,
    description: 'it should skip comment'
  }
]

describe('Tokenizer with config', () => {
  testCases.forEach(({ code, description }) => {
    test(description, () => {
      const tokenizer = new Tokenizer(config)
      tokenizer.parse(code)
      const tokenBuffer = tokenizer.getTokenBuffer()
      expect(tokenBuffer.toJSON()).toMatchSnapshot()
    })
  })
})