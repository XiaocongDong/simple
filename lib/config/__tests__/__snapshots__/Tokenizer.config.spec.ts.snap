// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tokenizer with config it should generate correct buffer 1`] = `
Array [
  Object {
    "range": Object {
      "end": Object {
        "column": 3,
        "line": 1,
      },
      "start": Object {
        "column": 1,
        "line": 1,
      },
    },
    "type": "LET",
    "value": "let",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 5,
        "line": 1,
      },
      "start": Object {
        "column": 5,
        "line": 1,
      },
    },
    "type": "IDENTIFIER",
    "value": "f",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 7,
        "line": 1,
      },
      "start": Object {
        "column": 7,
        "line": 1,
      },
    },
    "type": "ASSIGN",
    "value": "=",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 16,
        "line": 1,
      },
      "start": Object {
        "column": 9,
        "line": 1,
      },
    },
    "type": "FUNCTION",
    "value": "function",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 17,
        "line": 1,
      },
      "start": Object {
        "column": 17,
        "line": 1,
      },
    },
    "type": "LEFT_PAREN",
    "value": "(",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 18,
        "line": 1,
      },
      "start": Object {
        "column": 18,
        "line": 1,
      },
    },
    "type": "IDENTIFIER",
    "value": "x",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 19,
        "line": 1,
      },
      "start": Object {
        "column": 19,
        "line": 1,
      },
    },
    "type": "COMMA",
    "value": ",",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 21,
        "line": 1,
      },
      "start": Object {
        "column": 21,
        "line": 1,
      },
    },
    "type": "IDENTIFIER",
    "value": "y",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 22,
        "line": 1,
      },
      "start": Object {
        "column": 22,
        "line": 1,
      },
    },
    "type": "COMMA",
    "value": ",",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 24,
        "line": 1,
      },
      "start": Object {
        "column": 24,
        "line": 1,
      },
    },
    "type": "IDENTIFIER",
    "value": "z",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 25,
        "line": 1,
      },
      "start": Object {
        "column": 25,
        "line": 1,
      },
    },
    "type": "RIGHT_PAREN",
    "value": ")",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 27,
        "line": 1,
      },
      "start": Object {
        "column": 27,
        "line": 1,
      },
    },
    "type": "LEFT_CURLY_BRACE",
    "value": "{",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 12,
        "line": 2,
      },
      "start": Object {
        "column": 7,
        "line": 2,
      },
    },
    "type": "RETURN",
    "value": "return",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 14,
        "line": 2,
      },
      "start": Object {
        "column": 14,
        "line": 2,
      },
    },
    "type": "IDENTIFIER",
    "value": "x",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 15,
        "line": 2,
      },
      "start": Object {
        "column": 15,
        "line": 2,
      },
    },
    "type": "PLUS",
    "value": "+",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 16,
        "line": 2,
      },
      "start": Object {
        "column": 16,
        "line": 2,
      },
    },
    "type": "IDENTIFIER",
    "value": "y",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 17,
        "line": 2,
      },
      "start": Object {
        "column": 17,
        "line": 2,
      },
    },
    "type": "MULTIPLY",
    "value": "*",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 18,
        "line": 2,
      },
      "start": Object {
        "column": 18,
        "line": 2,
      },
    },
    "type": "IDENTIFIER",
    "value": "z",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 5,
        "line": 3,
      },
      "start": Object {
        "column": 5,
        "line": 3,
      },
    },
    "type": "RIGHT_CURLY_BRACE",
    "value": "}",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 6,
        "line": 3,
      },
      "start": Object {
        "column": 6,
        "line": 3,
      },
    },
    "type": "SEMICOLON",
    "value": ";",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 9,
        "line": 4,
      },
      "start": Object {
        "column": 5,
        "line": 4,
      },
    },
    "type": "CONST",
    "value": "const",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 11,
        "line": 4,
      },
      "start": Object {
        "column": 11,
        "line": 4,
      },
    },
    "type": "IDENTIFIER",
    "value": "a",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 13,
        "line": 4,
      },
      "start": Object {
        "column": 13,
        "line": 4,
      },
    },
    "type": "ASSIGN",
    "value": "=",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 15,
        "line": 4,
      },
      "start": Object {
        "column": 15,
        "line": 4,
      },
    },
    "type": "LEFT_SQUARE_BRACKET",
    "value": "[",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 16,
        "line": 4,
      },
      "start": Object {
        "column": 16,
        "line": 4,
      },
    },
    "type": "NUMBER_LITERAL",
    "value": "1",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 17,
        "line": 4,
      },
      "start": Object {
        "column": 17,
        "line": 4,
      },
    },
    "type": "COMMA",
    "value": ",",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 19,
        "line": 4,
      },
      "start": Object {
        "column": 19,
        "line": 4,
      },
    },
    "type": "NUMBER_LITERAL",
    "value": "2",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 20,
        "line": 4,
      },
      "start": Object {
        "column": 20,
        "line": 4,
      },
    },
    "type": "COMMA",
    "value": ",",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 22,
        "line": 4,
      },
      "start": Object {
        "column": 22,
        "line": 4,
      },
    },
    "type": "NUMBER_LITERAL",
    "value": "3",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 23,
        "line": 4,
      },
      "start": Object {
        "column": 23,
        "line": 4,
      },
    },
    "type": "COMMA",
    "value": ",",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 25,
        "line": 4,
      },
      "start": Object {
        "column": 25,
        "line": 4,
      },
    },
    "type": "LEFT_CURLY_BRACE",
    "value": "{",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 26,
        "line": 4,
      },
      "start": Object {
        "column": 26,
        "line": 4,
      },
    },
    "type": "NUMBER_LITERAL",
    "value": "1",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 27,
        "line": 4,
      },
      "start": Object {
        "column": 27,
        "line": 4,
      },
    },
    "type": "COMMA",
    "value": ",",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 29,
        "line": 4,
      },
      "start": Object {
        "column": 29,
        "line": 4,
      },
    },
    "type": "NUMBER_LITERAL",
    "value": "2",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 30,
        "line": 4,
      },
      "start": Object {
        "column": 30,
        "line": 4,
      },
    },
    "type": "COMMA",
    "value": ",",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 32,
        "line": 4,
      },
      "start": Object {
        "column": 32,
        "line": 4,
      },
    },
    "type": "NUMBER_LITERAL",
    "value": "3",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 33,
        "line": 4,
      },
      "start": Object {
        "column": 33,
        "line": 4,
      },
    },
    "type": "RIGHT_CURLY_BRACE",
    "value": "}",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 34,
        "line": 4,
      },
      "start": Object {
        "column": 34,
        "line": 4,
      },
    },
    "type": "COMMA",
    "value": ",",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 39,
        "line": 4,
      },
      "start": Object {
        "column": 36,
        "line": 4,
      },
    },
    "type": "STRING_LITERAL",
    "value": "'hi'",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 40,
        "line": 4,
      },
      "start": Object {
        "column": 40,
        "line": 4,
      },
    },
    "type": "RIGHT_SQUARE_BRACKET",
    "value": "]",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 41,
        "line": 4,
      },
      "start": Object {
        "column": 41,
        "line": 4,
      },
    },
    "type": "SEMICOLON",
    "value": ";",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 9,
        "line": 5,
      },
      "start": Object {
        "column": 5,
        "line": 5,
      },
    },
    "type": "CONST",
    "value": "const",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 11,
        "line": 5,
      },
      "start": Object {
        "column": 11,
        "line": 5,
      },
    },
    "type": "IDENTIFIER",
    "value": "o",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 13,
        "line": 5,
      },
      "start": Object {
        "column": 13,
        "line": 5,
      },
    },
    "type": "ASSIGN",
    "value": "=",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 15,
        "line": 5,
      },
      "start": Object {
        "column": 15,
        "line": 5,
      },
    },
    "type": "LEFT_CURLY_BRACE",
    "value": "{",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 10,
        "line": 6,
      },
      "start": Object {
        "column": 7,
        "line": 6,
      },
    },
    "type": "IDENTIFIER",
    "value": "name",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 11,
        "line": 6,
      },
      "start": Object {
        "column": 11,
        "line": 6,
      },
    },
    "type": "COLON",
    "value": ":",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 18,
        "line": 6,
      },
      "start": Object {
        "column": 13,
        "line": 6,
      },
    },
    "type": "STRING_LITERAL",
    "value": "'sean'",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 19,
        "line": 6,
      },
      "start": Object {
        "column": 19,
        "line": 6,
      },
    },
    "type": "COMMA",
    "value": ",",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 9,
        "line": 7,
      },
      "start": Object {
        "column": 7,
        "line": 7,
      },
    },
    "type": "IDENTIFIER",
    "value": "age",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 10,
        "line": 7,
      },
      "start": Object {
        "column": 10,
        "line": 7,
      },
    },
    "type": "COLON",
    "value": ":",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 13,
        "line": 7,
      },
      "start": Object {
        "column": 12,
        "line": 7,
      },
    },
    "type": "NUMBER_LITERAL",
    "value": "27",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 5,
        "line": 8,
      },
      "start": Object {
        "column": 5,
        "line": 8,
      },
    },
    "type": "RIGHT_CURLY_BRACE",
    "value": "}",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 6,
        "line": 8,
      },
      "start": Object {
        "column": 6,
        "line": 8,
      },
    },
    "type": "SEMICOLON",
    "value": ";",
  },
]
`;

exports[`Tokenizer with config it should skip comment 1`] = `
Array [
  Object {
    "range": Object {
      "end": Object {
        "column": 9,
        "line": 2,
      },
      "start": Object {
        "column": 7,
        "line": 2,
      },
    },
    "type": "LET",
    "value": "let",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 11,
        "line": 2,
      },
      "start": Object {
        "column": 11,
        "line": 2,
      },
    },
    "type": "IDENTIFIER",
    "value": "a",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 13,
        "line": 2,
      },
      "start": Object {
        "column": 13,
        "line": 2,
      },
    },
    "type": "ASSIGN",
    "value": "=",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 16,
        "line": 2,
      },
      "start": Object {
        "column": 15,
        "line": 2,
      },
    },
    "type": "NUMBER_LITERAL",
    "value": "10",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 17,
        "line": 2,
      },
      "start": Object {
        "column": 17,
        "line": 2,
      },
    },
    "type": "SEMICOLON",
    "value": ";",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 9,
        "line": 3,
      },
      "start": Object {
        "column": 7,
        "line": 3,
      },
    },
    "type": "LET",
    "value": "let",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 11,
        "line": 3,
      },
      "start": Object {
        "column": 11,
        "line": 3,
      },
    },
    "type": "IDENTIFIER",
    "value": "b",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 13,
        "line": 3,
      },
      "start": Object {
        "column": 13,
        "line": 3,
      },
    },
    "type": "ASSIGN",
    "value": "=",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 16,
        "line": 3,
      },
      "start": Object {
        "column": 15,
        "line": 3,
      },
    },
    "type": "NUMBER_LITERAL",
    "value": "40",
  },
  Object {
    "range": Object {
      "end": Object {
        "column": 17,
        "line": 3,
      },
      "start": Object {
        "column": 17,
        "line": 3,
      },
    },
    "type": "SEMICOLON",
    "value": ";",
  },
]
`;
